tests/test-util/src/url.spec.ts

```diff
@@ -24,4 +33,39 @@ describe("url", function () {
         expect(join("./", "bbb", "ccc")).to.equal("bbb/ccc");
         expect(join("/./", "bbb", "ccc")).to.equal("/bbb/ccc");
     });
+
+    it("posix dirname", function () {
+        expect(dirname("/aaa/bbb/ccc")).to.equal("/aaa/bbb");
+        expect(dirname("/aaa/bbb/ccc/")).to.equal("/aaa/bbb");
+        expect(dirname("/aaa/bbb/ccc/a")).to.equal("/aaa/bbb/ccc");
+    });
+
+    it("win32 join", function () {
+        expect(win32.join("aaa", "bbb", "ccc")).to.equal("aaa\\bbb\\ccc");
+        expect(win32.join()).to.equal("");
+        expect(win32.join("aaa")).to.equal("aaa");
+        expect(win32.join("\\aaa")).to.equal("\\aaa");
+        expect(win32.join("\\aaa", "bbb", "ccc")).to.equal("\\aaa\\bbb\\ccc");
+        expect(win32.join("\\aaa\\", "\\bbb\\", "\\ccc\\")).to.equal("\\aaa\\bbb\\ccc");
+        expect(win32.join("aaa\\", "\\bbb\\", "\\ccc\\")).to.equal("aaa\\bbb\\ccc");
+
+        expect(win32.join("aaa", "..\\bbb\\", ".\\ccc\\")).to.equal("bbb\\ccc");
+        expect(win32.join("\\aaa", "..\\bbb\\", ".\\ccc\\")).to.equal("\\bbb\\ccc");
+        expect(win32.join("\\aaa", ".\\bbb\\", ".\\ccc\\")).to.equal("\\aaa\\bbb\\ccc");
+        expect(win32.join("\\aaa", ".\\bbb\\", "..\\ccc\\")).to.equal("\\aaa\\ccc");
+        expect(win32.join("\\aaa", "..\\bbb\\", "..\\ccc\\")).to.equal("\\ccc");
+
+        expect(win32.join("\\aaa", ".\\bbb\\", "..\\..\\ccc\\")).to.equal("\\ccc");
+
+        expect(win32.join(".", "bbb", "ccc")).to.equal("bbb\\ccc");
+        expect(win32.join("\\.", "bbb", "ccc")).to.equal("\\bbb\\ccc");
+        expect(win32.join(".\\", "bbb", "ccc")).to.equal("bbb\\ccc");
+        expect(win32.join("\\.\\", "bbb", "ccc")).to.equal("\\bbb\\ccc");
+    });
+
+    it("win32 dirname", function () {
+        expect(win32.dirname("\\aaa\\bbb\\ccc")).to.equal("\\aaa\\bbb");
+        expect(win32.dirname("\\aaa\\bbb\\ccc\\")).to.equal("\\aaa\\bbb");
```

Comment:
And then, dependent upon the previous comment, should this equal "\\aaa\\bbb\\ccc\\" or possibly without the final dir separator?